### PR TITLE
Fix IPv6 host aliases for kubernetes

### LINKED ISSUE #2991: Invalid hostAliases.ip in IPv6-only cluster
### Component

agent

### Describe the bug

Trying to run pipeline with a service in IPv6-only cluster. Get error:

Screenshot 2023-12-22 141318

### System Info

```shell
Woodpecker next-619858e0
K3s 1.28.4
```

### Additional context

[woodpecker-agent.log](https://github.com/woodpecker-ci/woodpecker/files/13751825/woodpecker-agent.log)

```yaml
services:
  database:
    image: mysql
    environment:
      - MYSQL_DATABASE=test
      - MYSQL_ROOT_PASSWORD=example
    ports:
      - 3306/tcp
steps:
  get-version:
    image: ubuntu
    commands:
      - ( apt update && apt dist-upgrade -y && apt install -y mysql-client 2>&1 )> /dev/null
      - sleep 30s # need to wait for mysql-server init
      - echo 'SHOW VARIABLES LIKE "version"' | mysql -uroot -hdatabase test -pexample

```

### Validations

- [X] Read the [Contributing Guidelines](https://github.com/woodpecker-ci/woodpecker/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://woodpecker-ci.org/docs/intro).
- [X] Check that there isn't [already an issue](https://github.com/woodpecker-ci/woodpecker/issues) that reports the same bug to avoid creating a duplicate.
- [X] Checked that the bug isn't fixed in the `next` version already [https://woodpecker-ci.org/faq#which-version-of-woodpecker-should-i-use]
- [X] Check that this is a concrete bug. For Q&A join our [Discord Chat Server](https://discord.gg/fcMQqSMXJy) or the [Matrix room](https://matrix.to/#/#woodpecker:matrix.org).

### PR DESCRIPTION
**Pull request by zc-devs**
Closes #2991

[Tests](https://github.com/woodpecker-ci/woodpecker/pull/2993#issuecomment-1868048169)

### CODE REVIEW

File: pipeline/frontend/yaml/compiler/convert.go [hunk 1]
```diff
@@ -55,6 +55,16 @@ func (c *Compiler) createProcess(name string, container *yaml_types.Container, s
 		})
 	}
 
+	extraHosts := []backend_types.HostAlias{}
+	for _, extraHost := range container.ExtraHosts {
+		splitArr := strings.SplitN(extraHost, ":", 2)
```
Comment (lafriks) [inline pipeline/frontend/yaml/compiler/convert.go:60]: strings.Cut should be used instead

File: pipeline/frontend/yaml/compiler/convert.go [hunk 2]
```diff
@@ -55,6 +55,19 @@ func (c *Compiler) createProcess(name string, container *yaml_types.Container, s
...
+	for _, extraHost := range container.ExtraHosts {
+		name, ip, ok := strings.Cut(extraHost, ":")
+		if !ok {
+			return nil, fmt.Errorf("extra host %s is in wrong format", extraHost)
+		}
+		extraHost := backend_types.HostAlias{
+			Name: name,
+			IP:   ip,
+		}
+		extraHosts = append(extraHosts, extraHost)
```
Comment (6543) [inline pipeline/frontend/yaml/compiler/convert.go:68]:
```suggestion
		extraHosts = append(extraHosts, backend_types.HostAlias{
			Name: name,
			IP:   ip,
		})
```

**Review by 6543 — APPROVED** [review]

File: pipeline/frontend/yaml/compiler/convert.go [hunk 3]
```diff
@@ -55,6 +55,18 @@ func (c *Compiler) createProcess(name string, container *yaml_types.Container, s
 		})
 	}
 
+	extraHosts := []backend_types.HostAlias{}
```
Comment (6543) [inline pipeline/frontend/yaml/compiler/convert.go:58]:
```suggestion
	var extraHosts []backend_types.HostAlias
```

File: pipeline/frontend/yaml/compiler/convert.go [hunk 4]
```diff
@@ -55,6 +55,18 @@ func (c *Compiler) createProcess(name string, container *yaml_types.Container, s
 		})
 	}
 
+	var extraHosts []backend_types.HostAlias
```
Comment (lafriks) [inline pipeline/frontend/yaml/compiler/convert.go:58]: As the resulting array length is known it is better to initialize it either with enough capacity or with needed length initialy:
extraHosts := make([]..., len(container.ExtraHosts))

And the assign values already by loop index

**Review by lafriks — CHANGES_REQUESTED** [review]

File: pipeline/frontend/yaml/compiler/convert.go [hunk 5]
```diff
@@ -55,6 +55,18 @@ func (c *Compiler) createProcess(name string, container *yaml_types.Container, s
 		})
 	}
 
+	extraHosts := make([]backend_types.HostAlias, 0, len(container.ExtraHosts))
```
Comment (lafriks) [inline pipeline/frontend/yaml/compiler/convert.go:58]:
```suggestion
	extraHosts := make([]backend_types.HostAlias, len(container.ExtraHosts))
```

And then just extraHosts[i].Name = name and same for IP

This way array objects will be allocated at once thus making it more optimal

**Comment by 6543** [comment]
also added dedicated error :)

**Review by lafriks — APPROVED** [review]

**Comment by zc-devs** [comment]
Thank you, guys!

**Pull request merged**